feat: vola progressive integration

Tiavina-Andriamamivony opened the pull request.

Tiavina-Andriamamivony committed:
feat(psp): add Vola payment service provider integration

Tiavina-Andriamamivony committed:
chore: add VolaPsp configuration

Tiavina-Andriamamivony committed:
feat: Optional vola integration

Tiavina-Andriamamivony committed:
fix(test): fix MpbsVerificationService instantiation

Tiavina-Andriamamivony committed:
fix: format

Tiavina-Andriamamivony committed:
chore(vola): add vola api credentials

Tiavina-Andriamamivony committed:
chore: remove comments

Tiavina-Andriamamivony committed:
feat(test): VolaPspIT

Tiavina-Andriamamivony committed:
feat: complete VolaPsp

Tiavina-Andriamamivony committed:
chore: add vola integration base test

Tiavina-Andriamamivony committed:
fix: add missing VOLA API properties to test configuration

Tiavina-Andriamamivony committed:
chore: clean

Tiavina-Andriamamivony committed:
chore: remove unused auth components from Vola ApiClient

Tiavina-Andriamamivony committed:
chore: correct error handling

Tiavina-Andriamamivony committed:
chore: correct mapping and IT test

Tiavina-Andriamamivony committed:
chore: complete vola integration base test

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
chore(apiClient): rm condition

Tiavina-Andriamamivony committed:
chore: use default date format

Tiavina-Andriamamivony committed:
chore: rm unused date format

Tiavina-Andriamamivony committed:
chore: rm comments

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,96 @@
+  private final VolaClient volaClient;
+
+  @Override
+  public Mpbs create(Mpbs mpbs) {
```
create(pspType, pspId, email) like in VolaClient

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  @Override
+  public Mpbs get(Mpbs mpbs) {
```
get(pspType, pspId, email) like in VolaClient

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,96 @@
+      }
+      default ->
+          throw new UnsupportedOperationException(
+              "ORANGE_MONEY is the only MobileMoneyType supported by Vola currently");
```
No. "Unsupported mobileMoneyType: {mobileMoneyType}" so that we can see in the log what was the value used.

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  private Mpbs toMpbs(Mpbs mpbs, Payment volaPayment) {
+
```
Why the empty line...

hei-teacher commented on src/test/java/school/hei/haapi/conf/EnvConf.java:
```diff
@@ -17,12 +17,16 @@ void configureProperties(DynamicPropertyRegistry registry) {
     registry.add("spring.flyway.locations", () -> "classpath:/db/migration," + flywayTestdataPath);
     registry.add("CASDOOR_ORGANIZATION_NAME", () -> "dummy");
-    registry.add("CASDOOR_CERTIFICATE", () -> "dummy");
+    registry.add("CASDOOR_CERTIFICATE", () -> "ZHVtbXk=");
```
is this a secret!

hei-teacher commented on src/test/java/school/hei/haapi/conf/EnvConf.java:
```diff
@@ -17,12 +17,16 @@ void configureProperties(DynamicPropertyRegistry registry) {
+    registry.add(
+        "vola.api.url",
+        () -> "https://ypmoi24xu4kt5ts77p7te5uhme0uyxnv.lambda-url.eu-west-3.on.aws/");
+    registry.add("vola.api.key", () -> "ec48ed08-3a83-4b95-9f7e-ce5c07f541eb");
```
is this a secret!!

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -114,6 +116,60 @@ public List<MpbsVerification> verifyMobilePaymentAndSaveResult(List<Mpbs> pendin
+    List<Mpbs> unverifiedMpbs = new ArrayList<>();
+
+    for (Mpbs pendingMbps : pendingMpbsList) {
+      try {
```
Unclear

Tiavina-Andriamamivony committed:
chore: set dummy as default env

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

Tiavina-Andriamamivony committed:
chore: use  pspType, pspId and email like in VolaClient

Tiavina-Andriamamivony committed:
chore: enhance payment verification with multi-payment support

- Add fromVolaPayment method in transactionDetailsMapper for payment conversion
- Add getPayments method in PSP for multiple payments verification
- Add builder pattern in VolaPayment for flexible object creation
- Update VolaPspIt configuration for POJA application compatibility

Tiavina-Andriamamivony committed:
chore: enhance payment verification with multi-payment support

- Add fromVolaPayment method in transactionDetailsMapper for payment conversion
- Add getPayments method in PSP for multiple payments verification
- Add builder pattern in VolaPayment for flexible object creation
- Update VolaPspIt configuration for POJA application compatibility
-Update tests to accord to getPayments

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

Tiavina-Andriamamivony committed:
feat: vola updated integration

Vola api changed so I updated the base integration , now it include recovery process and multiple payment verification

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

Tiavina-Andriamamivony committed:
fix: add all arg constructor to PaymentInfo and update tests

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaClient.java:
```diff
@@ -0,0 +1,32 @@
+    return paymentControllerApi.getPayment(apiKey, email, pspType.toString(), pspId);
+  }
+
+  public List<Payment> getPayments(List<PaymentInfo> paymentInfos) {
```
Payment...Id?

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaClient.java:
```diff
@@ -0,0 +1,32 @@
+    return paymentControllerApi.createPayment(apiKey, email, pspType.toString(), pspId);
+  }
+
+  public Payment get(PspType pspType, String pspId, String email) {
```
get(PaymentId)

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,48 @@
+  @Override
+  public Payment create(PspType pspType, String pspId, String email) {
+    try {
+      log.info("Creating MPBS via Vola for student: {}", email);
```
log all payemntId

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,48 @@
+  @Override
+  public Payment get(PspType pspType, String pspId, String email) {
+    try {
+      log.info("Retrieving MPBS via Vola for student: {}", email);
```
all PaymentId

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,48 @@
+  @Override
+  public List<Payment> getPayments(List<PaymentInfo> paymentInfos) {
+    try {
+      return volaClient.getPayments(paymentInfos);
```
why log in other methods and not here. Be consistent: either you log all or log nothing

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -114,6 +121,66 @@ public List<MpbsVerification> verifyMobilePaymentAndSaveResult(List<Mpbs> pendin
   }
 
+  // Vola integration
+  public List<MpbsVerification> verifyMobilePaymentAndSaveResultWithVola(
```
Too long. Didn't even read

Tiavina-Andriamamivony committed:
chore: rename PaymentInfo to PaymentId and use it in getPayment

Tiavina-Andriamamivony committed:
chore(volaPsp): correct logging

Tiavina-Andriamamivony committed:
chore(MpbsVerificationService): refactor verifyMobilePaymentAndSaveResultWithVola method

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaClient.java:
```diff
@@ -0,0 +1,27 @@
+    this.paymentControllerApi = new PaymentControllerApi(apiClient);
+  }
+
+  public Payment get(PaymentId paymentId) {
```
getPayment

Tiavina-Andriamamivony committed:
chore(mpbs): improve verification resilience and eliminate side effects

Tiavina-Andriamamivony committed:
chore(mpbs): use vola in verifyMobilePaymentAndSaveResult

Tiavina-Andriamamivony committed:
chore(mpbs-verification): remove side effects on verifiedMpbs and return explicit results

hei-teacher commented on src/main/java/school/hei/haapi/model/PaymentProcessingResult.java:
```diff
@@ -0,0 +1,8 @@
+import school.hei.haapi.model.mpbs.Mpbs;
+import school.hei.haapi.model.mpbs.MpbsVerification;
+
+public record PaymentProcessingResult(
```
PayementVerificationResult, as you are separating verified from unverified

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+    return result.verifiedMpbs();
+  }
+
+  private PaymentProcessingResult processPaymentsIndividually(
```
Ban the word "process". We a more precise word. Like "verify".

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+
+  private PaymentProcessingResult processPaymentsIndividually(
+      List<Mpbs> pendingMpbsList, Map<String, Payment> paymentMap) {
+
```
useless blank line

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
-    unverifiedMobilePaymentHandler.accept(unverifiedMpbs);
-    return verifiedMpbs;
+    log.info(
+        "Processed all payments - Verified: {}, Unverified: {}",
```
Ban process everywhere

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+      Map<String, Payment> paymentMap,
+      List<MpbsVerification> verifiedMpbs,
+      List<Mpbs> unverifiedMpbs) {
+
```
useless

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+  }
+
+  private boolean isPaymentSuccessful(Payment payment) {
+    return payment.getVerificationStatus() == Payment.VerificationStatusEnum.SUCCEEDED;
```
static import

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+  private void logUnverifiedReason(Payment volaPayment, String pspId) {
+    if (volaPayment == null) {
+      log.info("No payment found in Vola for PSP ID: {}", pspId);
+    } else if (volaPayment.getVerificationStatus() == Payment.VerificationStatusEnum.FAILED) {
```
static import

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+
+      verifiedMpbs.add(
+          computeVerifiedMobilePayment.saveTheVerifiedMpbs(pendingMpbs, transactionDetails));
+
```
usless blank

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +46,175 @@ public class MpbsVerificationService {
+    }
+  }
+
+  private void processVerifiedPayment(
```
seems process here means save --> sometimes it means verify, sometimes it means save

which is why you should avoid generic word like "process". be super precise. rather be verbose but precise, rather than concise but imprecise

Tiavina-Andriamamivony committed:
chore: replace get method to getPayment

Tiavina-Andriamamivony committed:
chore: rename PaymentProcessingResult to PayementVerificationResult

Tiavina-Andriamamivony committed:
chore: ban process everywhere

Tiavina-Andriamamivony committed:
chore: use static import

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,47 @@
+  private final VolaClient volaClient;
+
+  @Override
+  public Payment getPayment(PaymentId paymentId) {
```
Nah... In one line: `return getPayments(List.of(paymentId))`

All exception treatments are already done in getPayments

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +48,172 @@ public class MpbsVerificationService {
-      }
+    Map<String, Payment> paymentMap;
+    try {
+      paymentMap = fetchAndMapVolaPayments(pendingMpbsCopy);
```
Map...ByWhat? ById? What is that String key? Be more precise.

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +48,172 @@ public class MpbsVerificationService {
+      return List.of();
+    }
+
+    var result = verifyPaymentsIndividually(pendingMpbsCopy, paymentMap);
```
Why is the `Individually` important? isn't verifyPayments enough?

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +48,172 @@ public class MpbsVerificationService {
-      } catch (NoRemainingAmountFee e) {
+    for (Mpbs pendingMpbs : pendingMpbsList) {
+      try {
+        verifySinglePayment(pendingMpbs, paymentMap, verifiedMpbs, unverifiedMpbs);
```
Naaah... This really needs a call at this point...

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +48,172 @@ public class MpbsVerificationService {
+    return new PaymentVerificationResult(verifiedMpbs, unverifiedMpbs);
+  }
+
+  private void verifySinglePayment(
```
Nah, verifyPayment should be enough. The fact that you need to precise `SinglePayment` is just because you do side effect on those lists which is bad...

Tiavina-Andriamamivony commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +48,172 @@ public class MpbsVerificationService {
+    return result.verifiedMpbs();
+  }
+
+  private PaymentVerificationResult verifyPaymentsIndividually(
```
Associate volaPaymlent with mpbs

Tiavina-Andriamamivony commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +48,172 @@ public class MpbsVerificationService {
+  }
+
+  private PaymentVerificationResult verifyPaymentsIndividually(
+      List<Mpbs> pendingMpbsList, Map<String, Payment> paymentMap) {
```
paymentsById

Tiavina-Andriamamivony replied:
Verify that mpbs are really in pending status, OR remove the "pending" it should work with any mpbs
(associateMpbs)

Tiavina-Andriamamivony committed:
chore: base version (without logs and error gestion)

Tiavina-Andriamamivony committed:
chore: delete PaymentVerificationResult

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/gen/client/model/mapper/VolaMapper.java:
```diff
@@ -0,0 +1,84 @@
+  public PaymentId.PspTypeEnum toPspTypeEnum(MobileMoneyType mobileMoneyType) {
+    switch (mobileMoneyType) {
+      case ORANGE_MONEY -> {
+        return PaymentId.PspTypeEnum.ORANGE_MONEY;
```
remove return and {}

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/gen/client/model/mapper/VolaMapper.java:
```diff
@@ -0,0 +1,84 @@
+
+  public PspPayment.PspTypeEnum toPspPaymentType(PspType pspType) {
+    switch (pspType) {
+      case ORANGE_MONEY -> {
```
too verbose

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/gen/client/model/mapper/VolaMapper.java:
```diff
@@ -0,0 +1,84 @@
+
+  public PspPayment.PspTypeEnum toPspPaymentType(PaymentId.PspTypeEnum pspType) {
+    switch (pspType) {
+      case ORANGE_MONEY -> {
```
...

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +43,59 @@ public class MpbsVerificationService {
-    }
+    var pendingMpbsCopy = List.copyOf(pendingMpbsList);
+
+    var paymentRetrievedFromVola =
```
payement_s_Retrieved...

Tiavina-Andriamamivony committed:
chore: genral refactor

Tiavina-Andriamamivony committed:
chore: use yoda condition to avoid NullPointerExcepetion

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,41 @@
+      var paymentRetrieved = volaClient.getPayments(paymentIds);
+
+      log.info("Successfully retrieved {} payments via Vola", paymentRetrieved.size());
+      log.debug("Retrieved payments: {}", paymentRetrieved);
```
Useless

hei-teacher commented on src/main/java/school/hei/haapi/model/psp/vola/api/VolaPsp.java:
```diff
@@ -0,0 +1,41 @@
+    try {
+      log.info("Retrieving {} payments via Vola - paymentIds: {}", paymentIds.size(), paymentIds);
+
+      var paymentRetrieved = volaClient.getPayments(paymentIds);
```
plural

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +44,80 @@ public class MpbsVerificationService {
+          e);
+      return List.of();
     }
+    log.debug(
```
no debug

hei-teacher commented on src/main/java/school/hei/haapi/service/MpbsVerificationService.java:
```diff
@@ -42,84 +44,80 @@ public class MpbsVerificationService {
+        paymentsRetrievedFromVola.stream()
+            .filter(payment -> SUCCEEDED.equals(payment.getVerificationStatus()))
+            .toList();
+    log.debug(
```
nah

hei-teacher commented on src/test/java/school/hei/haapi/model/psp/vola/VolaPspIT.java:
```diff
@@ -0,0 +1,41 @@
+
+  @Test
+  void read_succeeded_payment() {
+
```
useless

Tiavina-Andriamamivony committed:
test: update VolaPspIt  after refactor

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

Tiavina-Andriamamivony committed:
chore: clean

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

DyferHerioss committed:
feat(doc): add new endpoint to export all fees

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

Tiavina-Andriamamivony committed:
test: update import

Tiavina-Andriamamivony committed:
chore: update PaymentControllerApi

Tiavina-Andriamamivony committed:
Merge branch 'preprod' into merge-conflicts

Tiavina-Andriamamivony committed:
ci: revert CI

Tiavina-Andriamamivony committed:
chore: simplify volaPspIt

Tiavina-Andriamamivony committed:
ci: see loggs

Tiavina-Andriamamivony committed:
cd:; revert workflow

Tiavina-Andriamamivony committed:
wip:ci: resolve varenv visibility issue

Tiavina-Andriamamivony committed:
Merge branch 'preprod' into merge-conflicts

Tiavina-Andriamamivony committed:
wip:fix: varenvs visibility

Tiavina-Andriamamivony committed:
chore: clean
